src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -29,28 +31,34 @@ internal sealed partial class RemoteWorkspace : Workspace
         private readonly SemaphoreSlim _gate = new(initialCount: 1);
 
         /// <summary>
-        /// The last solution for the primary branch fetched from the client.
+        /// Mapping from solution-checksum to the solution computed for it.  This is used so that we can hold a solution
+        /// around as long as the checksum for it is being used in service of some feature operation (e.g.
+        /// classification).  As long as we're holding onto it, concurrent feature requests for the same solution
+        /// checksum can share the computation of that particular solution and avoid duplicated concurrent work.
         /// </summary>
-        private (Checksum checksum, Solution solution) _lastRequestedPrimaryBranchSolution;
+        private readonly Dictionary<Checksum, ReferenceCountedDisposable<LazySolution>> _solutionChecksumToLazySolution = new();
```

Comment:
Moved the dictionary up as i view it as the most important storage of our cached solutions.  The two fields below are now just fast accesses with an additional ref-count of the last 2 important solutions accessed.